src/cgroups/stats.rs

```diff
@@ -261,3 +264,128 @@ impl Default for HugeTlbStats {
+            continue;
+        }
+
+        let file_name = hugetlb_entry.file_name();
+        let file_name = file_name.to_str().unwrap();
+        if let Some(name_stripped) = file_name.strip_prefix("hugepages-") {
+            if let Some(size) = name_stripped.strip_suffix("kB") {
+                let size: u64 = size.parse()?;
+
+                let size_moniker = if size >= (1 << 20) {
+                    (size >> 20).to_string() + "GB"
+                } else if size >= (1 << 10) {
+                    (size >> 10).to_string() + "MB"
+                } else {
+                    size.to_string() + "KB"
+                };
+
+                sizes.push(size_moniker);
+            }
+        }
+    }
+
+    Ok(sizes)
+}
+
+pub fn parse_value(value: &str) -> Result<u64> {
+    value
+        .parse()
+        .with_context(|| format!("failed to parse {}", value))
+}
+
+pub fn parse_single_value(file_path: &Path) -> Result<u64> {
+    let value = common::read_cgroup_file(file_path)?;
+    let value = value.trim();
+    if value == "max" {
+        return Ok(u64::MAX);
+    }
+
+    value.parse().with_context(|| {
+        format!(
+            "failed to parse value {} from {}",
+            value,
+            file_path.display()
+        )
+    })
+}
+
+pub fn parse_flat_keyed_data(file_path: &Path) -> Result<HashMap<String, u64>> {
+    let mut stats = HashMap::new();
+    let keyed_data = common::read_cgroup_file(file_path)?;
+    for entry in keyed_data.lines() {
+        let entry_fields: Vec<&str> = entry.split_ascii_whitespace().collect();
+        if entry_fields.len() != 2 {
+            continue;
+        }
+
+        stats.insert(
+            entry_fields[0].to_owned(),
+            entry_fields[1].parse().with_context(|| {
+                format!(
+                    "failed to parse value {} from {}",
+                    entry_fields[0],
+                    file_path.display()
+                )
+            })?,
+        );
+    }
+
+    Ok(stats)
+}
+
+pub fn parse_nested_keyed_data(file_path: &Path) -> Result<HashMap<String, Vec<String>>> {
+    let mut stats: HashMap<String, Vec<String>> = HashMap::new();
+    let keyed_data = common::read_cgroup_file(file_path)?;
+    for entry in keyed_data.lines() {
+        let entry_fields: Vec<&str> = entry.split_ascii_whitespace().collect();
+        if entry_fields.len() < 2 {
+            continue;
+        }
+
+        stats.insert(
+            entry_fields[0].to_owned(),
+            entry_fields[1..]
+                .iter()
+                .copied()
+                .map(|p| p.to_owned())
+                .collect(),
+        );
+    }
+
+    Ok(stats)
+}
+
+pub fn parse_device_number(device: &str) -> Result<(u64, u64)> {
+    let numbers: Vec<&str> = device.split_terminator(':').collect();
+    if numbers.len() != 2 {
+        bail!("failed to parse device number {}", device);
+    }
+
+    Ok((numbers[0].parse()?, numbers[1].parse()?))
+}
```

Comment:
I want to ensure that the functions in this area that process strings are safe enough by testing.